sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -284,6 +298,19 @@ case class RangePartitioning(ordering: Seq[SortOrder], numPartitions: Int)
       }
     }
   }
+
+  override private[spark] def pruneInvalidAttribute(invalidAttr: Attribute): Partitioning = {
+    if (this.references.contains(invalidAttr)) {
+      val validExprs = this.children.takeWhile(!_.references.contains(invalidAttr))
+      if (validExprs.isEmpty) {
+        UnknownPartitioning(numPartitions)
+      } else {
+        RangePartitioning(validExprs, numPartitions)
```

Comment:
think about `RangePartitioning('a.ASC, 'b.ASC)` with output expression `'a as 'a1`.

It cannot satisfy `ClusteredDistribution('a1)`, but can still satisfy `OrderedDistribution('a1.ASC)`. I think the expected result should be `RangePartitioning('a1.ASC, 'b.ASC)`.

Reply:
Why doesn't it satisfy `ClusteredDistribution('a1)`? I don't agree with what you stated. If `b` is not in the output, it is useless to have it there. Moreover, when ordering we order always for the first attribute, then for the second, ... So if something is partitioned by `RangePartitioning('a1.ASC, 'b.ASC)`, it is also true that its partitioning is `RangePartitioning('a1.ASC, )`. So I think that in that case `RangePartitioning('a1.ASC)` is the right one.

Reply:
according to `RangePartitioning.satisfies0`, and the comments in `RangePartitioning` and `ClusteredDistribution`, `RangePartitioning('a.ASC, 'b.ASC)` does not satisfy `ClusteredDistribution('a)`.

Reply:
mmmh, I am not sure about what you mean referring to the classdoc of these 2 classes. I see nothing there about this. Anyway, I see that the implementation is done according what you state, but I do believe that it is wrong (or at least suboptimal if you prefer). If the data is partitioned by sorting it with `a.ASC, b.ASC`, it is definitely partitioned by sorting it with `a.ASC`. I think that `forall` should be an `exists`. There is also a (very minor) bug in the current implementation; try and running this test (it fails...):
```
  test("partitioning test") {
    val attr1 = AttributeReference("attr1", IntegerType)()
    val attr2 = AttributeReference("attr2", IntegerType)()
    val partitioning = RangePartitioning(Seq.empty, 10)
    val requiredDistribution = ClusteredDistribution(Seq(attr2, attr1), Some(10))
    assert(!partitioning.satisfies(requiredDistribution))
  }
```

Reply:
please carefully read the classdoc of `RangePartitioning` and `ClusteredDistribution`, and see what `RangePartitioning` guarantees and what `ClusteredDistribution` requires.
```
partition 1: (a=1, b=2), (a=1, b=3)
partition 2: (a=1, b=4), (a=1, b=5)
```
This data set is range partitioned by a,b, but not clustered by a.

Reply:
ah I see now, thanks. Let me update it then accordingly, thanks.